Enable migration from single-stack IPv4 to dual-stack IPv4, IPv6. (#11692)
gardener/gardener: docs/usage/networking/dual-stack-networking-migration.md

> ---

Reviewer: Please add a link to the new document to https://github.com/gardener/gardener/blob/master/docs/README.md.

Reviewer: Should we also link to this document from https://github.com/gardener/gardener/blob/master/docs/development/ipv6.md?

Reviewer: Clarified in offline discussion that `ipv6.md` will be updated at some later point in time.

title: Dual-stack network migration

> description: Migrate IPv4 shoots to dual-stack IPv,IPv6 network

Reviewer: ```suggestion
description: Migrate IPv4 shoots to dual-stack IPv4,IPv6 network
```

---

# Dual-Stack Network Migration

> This document provides a guide for migrating IPv4-based or IPv6-based Gardener shoot clusters to dual-stack networking (IPv4 and IPv6).

Reviewer: ```suggestion
This document provides a guide for migrating IPv4-only or IPv6-only Gardener shoot clusters to dual-stack networking (IPv4 and IPv6).

```

## Overview

> Dual-stack networking allows clusters to operate with both IPv4 and IPv6 protocols. This configuration is controlled via the `shoot.Spec.Networking.IPFamilies` field, which accepts the following values:

Reviewer: ```suggestion
Dual-stack networking allows clusters to operate with both IPv4 and IPv6 protocols. This configuration is controlled via the `spec.networking.ipFamilies` field, which accepts the following values:
```

- `[IPv4]`
- `[IPv6]`
- `[IPv4, IPv6]`
- `[IPv6, IPv4]`

### Key Considerations
- Adding a new protocol is only allowed as the second element in the array, ensuring the primary protocol remains unchanged.
- Migration involves multiple reconciliation runs to ensure a smooth transition without disruptions.

## Migration Process

### Step 1: Update Networking Configuration

> Modify the `shoot.Spec.Networking.IPFamilies` field to include the desired dual-stack configuration. For example, change `[IPv4]` to `[IPv4, IPv6]`.

Reviewer: ```suggestion
Modify the `spec.networking.ipFamilies` field to include the desired dual-stack configuration. For example, change `[IPv4]` to `[IPv4, IPv6]`.
```

### Step 2: Infrastructure Reconciliation

> Changing the `IPFamilies` field triggers an infrastructure reconciliation. This step applies necessary changes to the underlying infrastructure to support dual-stack networking.

Reviewer: ```suggestion
Changing the `ipFamilies` field triggers an infrastructure reconciliation. This step applies necessary changes to the underlying infrastructure to support dual-stack networking.
```

### Step 3: Control Plane Updates
Depending on the infrastructure, control plane components will be updated or reconfigured to support dual-stack networking.

### Step 4: Node Rollout

> Nodes must support the new network protocol. However, node rollout is not triggered automatically. It should be performed during a maintenance window to minimize disruptions. During shoot reconciliation, the system verifies if all nodes support dual-stack networking and updates the migration state accordingly.

Reviewer: Let's make it clear that this is a manual step, which will eventually happen automatically if nodes are for example rolled via kubernetes minor updates.
It might also make sense to add that cluster owners can check the status of this step via the shoot condition.

Reviewer: Thanks for the change, but shall we also mention the name of the node condition explicitly so that it is easier for users to check?

### Step 5: Final Reconciliation
Once all nodes are migrated, the remaining control plane components and the Container Network Interface (CNI) are configured for dual-stack networking. The migration constraint is removed at the end of this step.

## Post-Migration Behavior

After completing the migration:
- The shoot cluster supports dual-stack networking.
- New pods will receive IP addresses from both protocols.

> - Existing pods will only receive a second IP address upon restart.

Reviewer: ```suggestion
- New pods will receive IP addresses from both address families.
- Existing pods will only receive a second IP address upon recreation.
- If full dual-stack networking is required all pods need to be rolled.
- ```

---
title: Dual-stack network migration
description: Migrate IPv4 shoots to dual-stack IPv4,IPv6 network
---

# Dual-Stack Network Migration

This document provides a guide for migrating IPv4-only or IPv6-only Gardener shoot clusters to dual-stack networking (IPv4 and IPv6).

> ## Overview

Reviewer 2: ```suggestion

## Overview
```

Author: done

> Dual-stack networking allows clusters to operate with both IPv4 and IPv6 protocols. This configuration is controlled via the `spec.networking.ipFamilies` field, which accepts the following values:

Reviewer 2: ```suggestion
Dual-stack networking allows clusters to operate with both IPv4 and IPv6 protocols. This configuration is controlled via the `spec.networking.ipFamilies` field which accepts the following values:
```

Author: The comma before "which" serves to separate the main clause from the non-restrictive relative clause. Non-restrictive clauses add extra information that is not essential to the main point of the sentence. Omitting the comma would make the sentence grammatically incorrect according to traditional punctuation rules.

- `[IPv4]`
- `[IPv6]`
- `[IPv4, IPv6]`
- `[IPv6, IPv4]`

> ### Key Considerations

Reviewer 2: ```suggestion
### Key Considerations

```

Author: done

- Adding a new protocol is only allowed as the second element in the array, ensuring the primary protocol remains unchanged.
- Migration involves multiple reconciliation runs to ensure a smooth transition without disruptions.

## Migration Process

> ### Step 1: Update Networking Configuration

Reviewer 2: ```suggestion
### Step 1: Update Networking Configuration

```

Author: done

Modify the `spec.networking.ipFamilies` field to include the desired dual-stack configuration. For example, change `[IPv4]` to `[IPv4, IPv6]`.

> ### Step 2: Infrastructure Reconciliation

Reviewer 2: ```suggestion
### Step 2: Infrastructure Reconciliation

```

Author: done

Changing the `ipFamilies` field triggers an infrastructure reconciliation. This step applies necessary changes to the underlying infrastructure to support dual-stack networking.

> ### Step 3: Control Plane Updates

Reviewer 2: ```suggestion
### Step 3: Control Plane Updates

```

Author: done

Depending on the infrastructure, control plane components will be updated or reconfigured to support dual-stack networking.

### Step 4: Node Rollout

Nodes must support the new network protocol. However, node rollout is a manual step and is not triggered automatically. It should be performed during a maintenance window to minimize disruptions. Over time, this step may occur automatically, for example, during Kubernetes minor version updates that involve node replacements.

Cluster owners can monitor the progress of this step by checking the `DualStackNodesMigrationReady` constraint in the shoot status. During shoot reconciliation, the system verifies if all nodes support dual-stack networking and updates the migration state accordingly.

> ### Step 5: Final Reconciliation

Reviewer 2: ```suggestion
### Step 5: Final Reconciliation

```

Author: done

Once all nodes are migrated, the remaining control plane components and the Container Network Interface (CNI) are configured for dual-stack networking. The migration constraint is removed at the end of this step.

## Post-Migration Behavior

After completing the migration:
- The shoot cluster supports dual-stack networking.

>  New pods will receive IP addresses from both address families.

Reviewer 2: ```suggestion
- New pods will receive IP addresses from both address families.
```

- Existing pods will only receive a second IP address upon recreation.
- If full dual-stack networking is required all pods need to be rolled.

> 

Reviewer 2: Delete empty lines
